packages/koa-shopify-auth/src/auth/client/storage-access-helper.ts

```diff
@@ -0,0 +1,151 @@
+const storageAccessHelper = `(function() {
+      var ACCESS_GRANTED_STATUS = 'storage_access_granted';
+      var ACCESS_DENIED_STATUS = 'storage_access_denied';
+
+      function StorageAccessHelper(redirectData) {
+        this.redirectData = redirectData;
+      }
+
+      StorageAccessHelper.prototype.setNormalizedLink = function(storageAccessStatus) {
+        return storageAccessStatus === ACCESS_GRANTED_STATUS ? this.redirectData.hasStorageAccessUrl : this.redirectData.doesNotHaveStorageAccessUrl;
+      }
+
+      StorageAccessHelper.prototype.redirectToAppTLD = function(storageAccessStatus) {
+        var normalizedLink = document.createElement('a');
+
+        normalizedLink.href = this.setNormalizedLink(storageAccessStatus);
+
+        data = JSON.stringify({
+          message: 'Shopify.API.remoteRedirect',
+          data: {
+            location: normalizedLink.href,
+          }
+        });
+        window.parent.postMessage(data, this.redirectData.myshopifyUrl);
+      }
+
+      StorageAccessHelper.prototype.redirectToAppsIndex = function() {
+        window.parent.location.href = this.redirectData.myshopifyUrl + '/admin/apps';
+      }
+
+      StorageAccessHelper.prototype.redirectToAppTargetUrl = function() {
+        window.location.href = this.redirectData.appTargetUrl;
+      }
+
+      StorageAccessHelper.prototype.sameSiteNoneIncompatible = function(ua) {
+        return ua.includes("iPhone OS 12_") || ua.includes("iPad; CPU OS 12_") || //iOS 12
+        (ua.includes("UCBrowser/")
+            ? this.isOlderUcBrowser(ua) //UC Browser < 12.13.2
+            : (ua.includes("Chrome/5") || ua.includes("Chrome/6"))) ||
+        ua.includes("Chromium/5") || ua.includes("Chromium/6") ||
+        (ua.includes(" OS X 10_14_") &&
+            ((ua.includes("Version/") && ua.includes("Safari")) || //Safari on MacOS 10.14
+            ua.endsWith("(KHTML, like Gecko)"))); //Web view on MacOS 10.14
+      }
+
+      StorageAccessHelper.prototype.isOlderUcBrowser = function(ua) {
+        var match = ua.match(/UCBrowser\\/(\\d+)\\.(\\d+)\\.(\\d+)\\./);
+        if (!match) return false;
+        var major = parseInt(match[1]);
+        var minor = parseInt(match[2]);
+        var build = parseInt(match[3]);
+        if (major != 12) return major < 12;
+        if (minor != 13) return minor < 13;
+        return build < 2;
+      }
+
+      StorageAccessHelper.prototype.setCookie = function(value) {
+        if(!this.sameSiteNoneIncompatible(navigator.userAgent)) {
+          value += '; secure; SameSite=None'
+        }
+        document.cookie = value;
+      }
+
+      StorageAccessHelper.prototype.grantedStorageAccess = function() {
+        try {
+          sessionStorage.setItem('shopify.granted_storage_access', true);
+          this.setCookie('shopify.granted_storage_access=true');
+          if (!document.cookie) {
+            throw 'Cannot set third-party cookie.'
+          }
+          this.redirectToAppTargetUrl();
```

Comment:
Hi @ayronshopify @ismail-syed @TheMallen 

I'm a Shopify partner who's trying to understand this authorization process and I'm confused about this line and was hoping to get some clarification.

Why do we redirect the user to `this.redirectData.appTargetUrl` rather than `/auth`, which is the [`oAuthStartPath`](https://github.com/Shopify/quilt/blob/master/packages/koa-shopify-auth/src/auth/index.ts#L42) in the [`createShopifyAuth`](https://github.com/Shopify/quilt/blob/master/packages/koa-shopify-auth/src/auth/index.ts#L31) function?

The logic that checks the `shopify.granted_storage_access` cookie is associated with the `/auth` path and so it seems to me that we should be redirecting the user to `/auth`

Thank you!

Reply:
Hi! Good question. I think the reason is that we may have already completed OAuth. If we're already completed OAuth, and now we also have storage access, then the app is ready to load, so it doesn't make sense to do OAuth again.

If we haven't completed OAuth yet, then when we hit the app target URL we'll realize we need to do OAuth, and then redirect to OAuth anyway, so that case is also covered.

Does that make sense?